File: chazm-api/src/main/java/runtimemodels/chazm/api/relation/manager/ContainsManager.java
package runtimemodels.chazm.api.relation.manager;

import runtimemodels.chazm.api.entity.Characteristic;
import runtimemodels.chazm.api.entity.Role;
import runtimemodels.chazm.api.id.CharacteristicId;
import runtimemodels.chazm.api.id.RoleId;
import runtimemodels.chazm.api.id.UniqueId;
import runtimemodels.chazm.api.relation.Contains;

import java.util.Set;

/**
 * The {@linkplain ContainsManager} interface defines the APIs for managing {@linkplain Contains}.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public interface ContainsManager {
    /**
     * Creates a contains relation between a {@linkplain Role} and a {@linkplain Characteristic} in this {@linkplain ContainsManager}.
     *
     * @param roleId           the {@linkplain UniqueId} that represents the {@linkplain Role}.
     * @param characteristicId the {@linkplain UniqueId} that represents the {@linkplain Characteristic}.
     * @param value            the <code>double</code> value.
     */
    void addContains(RoleId roleId, CharacteristicId characteristicId, double value);

    /**
     * Returns a set of {@linkplain Characteristic}s that is contained by a {@linkplain Role} from this {@linkplain ContainsManager}.
     *
     * @param id the {@linkplain UniqueId} that represents the {@linkplain Role}.
     * @return a set of {@linkplain Characteristic}s.
     */
    Set<Characteristic> getContains(RoleId id);

    /**
     * Returns a set of {@linkplain Role}s that contains a {@linkplain Characteristic} from this {@linkplain ContainsManager}.
     *
     * @param id the {@linkplain UniqueId} that represents the {@linkplain Characteristic}.
     * @return a set of {@linkplain Role}s.
     */
    Set<Role> getContainedBy(CharacteristicId id);

    /**
     * Return a {@linkplain Double} value (if it exists) of the contains relation between a {@linkplain Role} and a {@linkplain Characteristic} from this
     * {@linkplain ContainsManager}.
     *
     * @param roleId           the {@linkplain UniqueId} that represents the {@linkplain Role}.
     * @param characteristicId the {@linkplain UniqueId} that represents the {@linkplain Characteristic}.
     * @return the {@linkplain Double} value if it exists, <code>null</code> otherwise.
     */
    Double getContainsValue(RoleId roleId, CharacteristicId characteristicId);

    /**
     * Sets a new <code>double</code> value of the contains relation between a {@linkplain Role} and a {@linkplain Characteristic} in this
     * {@linkplain ContainsManager}.
     *
     * @param roleId           the {@linkplain UniqueId} that represents the {@linkplain Role}.
     * @param characteristicId the {@linkplain UniqueId} that represents the {@linkplain Characteristic}.
     * @param value            the new <code>double</code> value.
     */
    void setContainsValue(RoleId roleId, CharacteristicId characteristicId, double value);

    /**
     * Removes a contains relation between a {@linkplain Role} and a {@linkplain Characteristic} from this {@linkplain ContainsManager}.
     *
     * @param roleId           the {@linkplain UniqueId} of the {@linkplain Role} for the {@linkplain Contains}.
     * @param characteristicId the {@linkplain UniqueId} of the {@linkplain Characteristic} for the {@linkplain Contains}.
     */
    void removeContains(RoleId roleId, CharacteristicId characteristicId);

    /**
     * Removes all contains relations from this {@linkplain ContainsManager}.
     */
    void removeAllContains();
}

File: chazm-api/src/main/kotlin/runtimemodels/chazm/api/organization/ContainsManager.kt
package runtimemodels.chazm.api.organization

import runtimemodels.chazm.api.entity.Characteristic
import runtimemodels.chazm.api.entity.Role
import runtimemodels.chazm.api.id.CharacteristicId
import runtimemodels.chazm.api.id.RoleId
import runtimemodels.chazm.api.relation.Contains

/**
 * The [ContainsManager] interface defines the APIs for managing a set of [Contains] relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
interface ContainsManager : Map<RoleId, Map<CharacteristicId, Contains>> {
    /**
     * Adds a [Contains] relation between a [Role] and a [Characteristic] to this [ContainsManager].
     *
     * @param contains           the [Contains] relation to add.
     */
    fun add(contains: Contains)

    /**
     * Returns the [Contains] relation between an [Role] and an [Characteristic] from this [HasManager].
     *
     * @param roleId           the [RoleId] that represents the [Role].
     * @param characteristicId the [CharacteristicId] that represents the [Characteristic].
     * @return the [Contains] relation if it exists, `null` otherwise.
     */
    operator fun get(roleId: RoleId, characteristicId: CharacteristicId): Contains?

    /**
     * Returns a [Map] of [Characteristic]s that are contained by a [Role] from this [ContainsManager].
     *
     * @param key the [RoleId] that represents the [Role].
     * @return a [Map] of [Characteristic]s.
     */
    override operator fun get(key: RoleId): Map<CharacteristicId, Contains>

    /**
     * Returns a [Map] of [Role]s that contains a [Characteristic] from this [ContainsManager].
     *
     * @param id the [CharacteristicId] that represents the [Characteristic].
     * @return a [Map] of [Role]s.
     */
    operator fun get(id: CharacteristicId): Map<RoleId, Contains>

    /**
     * Removes a [Contains] relation between a [Role] and a [Characteristic] from this [ContainsManager].
     *
     * @param roleId           the [RoleId] that represents the [Role].
     * @param characteristicId the [CharacteristicId] that represents the [Characteristic].
     * @return the [Characteristic] that was removed.
     */
    fun remove(roleId: RoleId, characteristicId: CharacteristicId): Contains

    /**
     * Removes all [Contains] relations associated to a [Role] from this [ContainsManager].
     *
     * @param id the [RoleId] that represents the [Role].
     */
    fun remove(id: RoleId)

    /**
     * Removes all [Contains] relations associated to a [Characteristic] from this [ContainsManager].
     *
     * @param id the [CharacteristicId] that represents the [Characteristic].
     */
    fun remove(id: CharacteristicId)
}
